Test/civl/rwlock.bpl

```diff
@@ -0,0 +1,152 @@
+// RUN: %parallel-boogie "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// In this example we show how to model a reader-writer lock.
+
+type {:linear "tid"} Tid;
+
+type {:datatype} OptionTid;
+function {:constructor} None(): OptionTid;
+function {:constructor} Some(tid: Tid): OptionTid;
+
+// An `RwLock` allows either a number of `readers` or at most one `writer` at
+// any point in time.
+type {:datatype} RwLock;
+function {:constructor} RwLock(writer: OptionTid, readers: [Tid]bool): RwLock;
+
+// We want to obtain the following mover types.
+// * Acquiring a read or write lock is a right mover.
+// * Releasing a read or write lock is a left mover.
+// * Actions that read while holding a read lock and actions that write while
+//   holding a write lock are both (left and right) movers.
+
+////////////////////////////////////////////////////////////////////////////////
+
+var {:layer 0, 1} rwlock: RwLock;
+
+// Acquiring a read lock is possible if there is no writer, and has the effect
+// of adding us to `readers`.
+procedure {:right}{:layer 1, 1} atomic_acquire_read({:linear "tid"} tid: Tid)
+modifies rwlock;
+{
+    assume writer#RwLock(rwlock) == None();
+    rwlock := RwLock(writer#RwLock(rwlock), readers#RwLock(rwlock)[tid := true]);
+}
+
+// Acquiring a write lock is possbile if there is no other writer and no reader,
+// and has the effect of storing us as the `writer`.
+procedure {:right}{:layer 1, 1} atomic_acquire_write({:linear "tid"} tid: Tid)
+modifies rwlock;
+{
+    assume writer#RwLock(rwlock) == None();
+    assume (forall tid: Tid :: !readers#RwLock(rwlock)[tid]);
```

Comment:
Why not use the generalized array theory?

Reply:
Done.